src/common/walFilter/walFilter.c

```diff
@@ -0,0 +1,706 @@
+    {
+        // We have an incomplete record at the end
+        if (this->currentStep != noStep)
+        {
+            const size_t size_on_page = this->gotLen;
+
+            // if xl_info and xl_rmid of the header is in current file then read end of record from next file if it exits
+            if (this->gotLen >= offsetof(XLogRecord, xl_rmid) + SIZE_OF_STRUCT_MEMBER(XLogRecord, xl_rmid))
+            {
+                getEndOfRecord(this);
+                filterRecord(this);
+            }
+
+            bufCatC(output, (const unsigned char *) this->record, 0, size_on_page);
+        }
+        this->done = true;
+        goto end;
+    }
+
+    if (this->isBegin)
+    {
+        this->isBegin = false;
+
+        readPage(this, input, noStep);
+        this->recPtr = this->currentHeader->xlp_pageaddr;
+        if (this->currentHeader->xlp_info & XLP_FIRST_IS_CONTRECORD && !(this->currentHeader->xlp_info & XLP_FIRST_IS_OVERWRITE_CONTRECORD))
+        {
+            if (readBeginOfRecord(this))
+            {
+                // Remember how much we read from the prev file in order to skip this size when writing.
+                const size_t offset = this->gotLen;
+                this->inputOffset = 0;
+                lstClearFast(this->headers);
+                if (readRecord(this, input) == ReadRecordNeedBuffer)
+                {
+                    // Since we are at the very beginning of the file, let's assume that the current input buffer is enough to fully
+                    // read this record.
+                    THROW_FMT(FormatError, "%s - record is too big", strZ(pgLsnToStr(this->recPtr)));
+                }
+                filterRecord(this);
+
+                ASSERT(offset <= 8);
+                memmove(this->record, ((char *) this->record) + offset, this->record->xl_tot_len - offset);
+                this->gotLen -= offset;
+
+                ASSERT(this->recPtr == this->currentHeader->xlp_pageaddr);
+                writeRecord(this, output);
+                this->sameInput = true;
+                lstClearFast(this->headers);
+                goto end;
+            } // else
+
+            this->inputOffset = 0;
+            readPage(this, input, noStep);
+            ASSERT(this->recPtr == this->currentHeader->xlp_pageaddr);
+            bufCatC(output, (const unsigned char *) this->currentHeader, 0, SizeOfXLogLongPHD);
+
+            this->recPtr += SizeOfXLogLongPHD;
+            lstClearFast(this->headers);
+
+            // The header that needs to be modified is in another file or not exists. Just copy it as is.
+            bufCatC(output, this->page, this->pageOffset, MAXALIGN(this->currentHeader->xlp_rem_len));
+
+            this->pageOffset += MAXALIGN(this->currentHeader->xlp_rem_len);
+            this->recPtr += MAXALIGN(this->currentHeader->xlp_rem_len);
+        }
+    }
+
+    // When meeting wal switch record, we write the rest of the file as is.
+    if (this->isSwitchWal)
```

Comment:
Can the `isSwitchWal` flag be set when calling `readRecord` from `getEndOfRecord` or `readBeginOfRecord`?

Reply:
No. If the next file (relative to the one being read) exists, then there should not be a switch wal record in it.

Reply:
switch wal - is this the last record we will read from the file?

Reply:
If this is the last file in the archive.